src/V3SplitVar.cpp

```diff
@@ -0,0 +1,250 @@
+// redistribute it and/or modify it under the terms of either the GNU
+// Lesser General Public License Version 3 or the Perl Artistic License
+// Version 2.0.
+//
+// Verilator is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+//*************************************************************************
+// V3SplitVar divides a variable into multiple variables to avoid UNOPTFLAT warning
+// and get better perfomance.
+// Variables to be splitted must be marked by /*verilator split_var*/ pragma.
+// There are sveral kinds of data types that may cause the warning.
+// 1) Unpacked arrays
+// 2) Packed arrays
+// 3) Unpacked structs
+// 4) Packed structs
+// 5) Bitfields within a signal. (Especially Verilog code predating structs/2D arrays.)
+// 2-5 above are treated as bitfields in verilator.
+//
+// What this pass does looks as below.
+//
+// <pre>
+// logic [1:0] unpcked_array_var[0:3]; /*verilator split_var*/
+// always_comb begin
+//    unpacked_array_var[1] = unpacked_array_var[0]; // UNOPTFLAT warning
+//    unpacked_array_var[2] = unpacked_array_var[1];
+//    unpacked_array_var[3] = unpacked_array_var[2];
+// end
+// </pre>
+//
+// is converted to
+//
+// <pre>
+// logic [1:0] unpcked_array_var0;
+// logic [1:0] unpcked_array_var1;
+// logic [1:0] unpcked_array_var2;
+// logic [1:0] unpcked_array_var3;
+// always_comb begin
+//    unpacked_array_var1 = unpacked_array_var0;
+//    unpacked_array_var2 = unpacked_array_var1;
+//    unpacked_array_var3 = unpacked_array_var2;
+// end
+// </pre>
+//
+//
+// Limitations: (planned to be resolved)
+// - Only 1D unpacked array can be splitted
+// - Splitting 2) - 5) will be supported later
+//
+//*************************************************************************
+
+#include "config_build.h"
+#include "verilatedos.h"
+
+#include "V3Ast.h"
+#include "V3Const.h"
+#include "V3Global.h"
+#include "V3SplitVar.h"
+#include "V3Stats.h"
+
+#include <vector>
+#include VL_INCLUDE_UNORDERED_MAP
+#include VL_INCLUDE_UNORDERED_SET
+
+namespace {
+//######################################################################
+//
+
+class SplitVarVisitor : public AstNVisitor {
+    AstNodeModule* m_modp;  // current module
+    AstVar* m_lastVar;      // the most recently declared variable
+    // key:variable to be splitted. value:location where the variable is referenced.
+    vl_unordered_map<AstVar*, std::vector<AstArraySel*> > m_refs;
+    virtual void visit(AstNode* nodep) VL_OVERRIDE;
+    virtual void visit(AstNodeModule* nodep) VL_OVERRIDE;
+    virtual void visit(AstVar* nodep) VL_OVERRIDE;
+    virtual void visit(AstPragma* pragmap) VL_OVERRIDE;
+    virtual void visit(AstArraySel* nodep) VL_OVERRIDE;
+
+public:
+    SplitVarVisitor(AstNodeModule* modp);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Class 'SplitVarVisitor' has a constructor with 1 argument that is not explicit.](https://app.codacy.com/app/wsnyder/verilator/pullRequest?prid=4724658)